[WIP] Try packing hot type flags in the `Ty` (pointer)

WaffleLapkin opened the pull request:
r? @ghost

tmandry committed:
coverage: Don't underflow column number

I noticed this when running coverage on a debug build of rustc. There
may be other places that do this but I'm just fixing the one I hit.

zhaixiaojuan committed:
Add loongarch64 asm! support

pat-nel87 committed:
black_box doc corrections - Issue #107957

clubby789 committed:
Add deny lint to prevent untranslatable diagnostics using static strings

pat-nel87 committed:
black_box hint - Adjust for improved readability

rytheo committed:
Add support for allocators in LinkedList

flba-eb committed:
Replace `yes` command by `while-echo`

The `yes` command is not available on all platforms.

WaffleLapkin committed:
Try to improve tidy errors on TODOs

pat-nel87 committed:
Merge branch 'master' into Issue-107957-black_box_docs

jsoref committed:
spelling: github

Signed-off-by: Josh Soref <2119212+jsoref@users.noreply.github.com>

jsoref committed:
spelling: typographical

Signed-off-by: Josh Soref <2119212+jsoref@users.noreply.github.com>

clubby789 committed:
Fix static string lints

jsoref committed:
Spelling library/

* advance
* aligned
* borrowed
* calculate
* debugable
* debuggable
* declarations
* desugaring
* documentation
* enclave
* ignorable
* initialized
* iterator
* kaboom
* monomorphization
* nonexistent
* optimizer
* panicking
* process
* reentrant
* rustonomicon
* the
* uninitialized

Signed-off-by: Josh Soref <2119212+jsoref@users.noreply.github.com>

tgross35 committed:
Update documentation wording on path 'try_exists' functions

obeis committed:
Migrate `rustc_hir_analysis` to session diagnostic

Part 4: Finishing `check/mod.rs` file

aDotInTheVoid committed:
Move test from rustdoc-ui to rustdoc-json

calebcartwright committed:
docs(style): add more let-else examples

Co-authored-by: Michael Goulet <michael@errs.io>

WaffleLapkin committed:
Allow using `Interned<..>` in tagged pointers

WaffleLapkin committed:
Prepare for using tagged pointers in `Ty`

WaffleLapkin committed:
tmp

WaffleLapkin committed:
tmp2

kylematsuda committed:
change usages of explicit_item_bounds to bound_explicit_item_bounds

kylematsuda committed:
add EarlyBinder to output of explicit_item_bounds; replace bound_explicit_item_bounds usages; remove bound_explicit_item_bounds query

kylematsuda committed:
add subst_identity_iter and subst_identity_iter_copied methods on EarlyBinder; use this to simplify some EarlyBinder noise around explicit_item_bounds calls

jsoref committed:
rewrite: long_line_flushed description

Signed-off-by: Josh Soref <2119212+jsoref@users.noreply.github.com>

jsoref committed:
rewrite: line_long_tail_not_flushed description

Signed-off-by: Josh Soref <2119212+jsoref@users.noreply.github.com>

compiler-errors committed:
Suggest deref on comparison binop RHS even if type is not Copy

ComputerDruid committed:
Add definitions for riscv64gc-unknown-fuchsia

tomaka committed:
Fix `std` compilation error for wasi+atomics

infdahai committed:
add cfg SPEC declaration.

use name[=value] syntax.

Signed-off-by: clundro <859287553@qq.com>

WaffleLapkin committed:
Add `impl_tag!` macro to implement `Tag` for tagged pointer easily

kylematsuda committed:
Changes from review

WaffleLapkin committed:
Use `impl Tag for $T` syntax for `impl_tag!`

WaffleLapkin committed:
Remove confusing comment

WaffleLapkin committed:
Simplify `bits_for_tags` impl

saethlin committed:
Rewrite MemDecoder around pointers not a slice

GuillaumeGomez committed:
Add new rustdoc book chapter to describe in-doc settings

arlosi committed:
Fix no_global_oom_handling build

`provide_sorted_batch` in core is incorrectly marked with
`#[cfg(not(no_global_oom_handling))]` which prevents core
from building with the cfg enabled.

Nothing in core allocates memory including this function, so
the `cfg` gate is incorrect.

compiler-errors committed:
vars are ?

compiler-errors committed:
Consider polarity in new solver

heiher committed:
Update dependencies

To support LoongArch.

Changes:
  Updating errno v0.2.8 -> v0.3.1
  Updating is-terminal v0.4.4 -> v0.4.7
  Updating linux-raw-sys v0.1.4 -> v0.3.4
  Updating rustix v0.36.5 -> v0.37.7
  Updating terminal_size v0.2.3 -> v0.2.6
    Adding windows-sys v0.48.0

The changes are generated by:

  cargo update -p is-terminal -p terminal_size

cjgillot committed:
Do not bother optimizing impossible functions.

WaffleLapkin committed:
Switch `impl_tag!` from explicit tags to `${index()}`

compiler-errors committed:
Negative coherence test

cjgillot committed:
Fully clear the body.

dtolnay committed:
Fix ICE on --print=... i/o errors

cuviper committed:
Revert "Download the GCC sources insecurely"

This reverts commit 3da037f82988b8b3aca2ce13c5c81ba975923cab.

rylev committed:
Use the standard macOS CI runner

bors committed:
Auto merge of #101069 - zhaixiaojuan:loongarch64-inline-asm, r=Amanieu

Add loongarch64 asm! support

WaffleLapkin committed:
rebase fix

WaffleLapkin committed:
debug_assrt

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4796873523/jobs/8533149279) [(plain)](https://github.com/rust-lang/rust/commit/b2c4ed4f306ce1a1e7cc888be5688aa5568ce9a4/checks/13004157871/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_ast_passes v0.0.0 (/checkout/compiler/rustc_ast_passes)
error: hidden lifetime parameters in types are deprecated
    --> compiler/rustc_middle/src/ty/context.rs:1296:49
     |
1296 |             .contains_pointer_to::<InternedInSet<WithCachedTypeInfo<TyKind<'a>>>>(&InternedInSet(&*self.0.pointer().0))
     |                                    -------------^------------------------------- expected lifetime parameter
     |
     = note: `-D elided-lifetimes-in-paths` implied by `-D warnings`
help: indicate the anonymous lifetime
     |
1296 |             .contains_pointer_to::<InternedInSet<'_, WithCachedTypeInfo<TyKind<'a>>>>(&InternedInSet(&*self.0.pointer().0))

   Compiling rustc_expand v0.0.0 (/checkout/compiler/rustc_expand)
   Compiling rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error: could not compile `rustc_middle` due to previous error
```

</details>

WaffleLapkin committed:
unhide lifetime param

bors committed:
Auto merge of #103093 - rytheo:linked-list-alloc-api, r=Mark-Simulacrum

Add support for allocators in `LinkedList`

Allows `LinkedList` to use a custom allocator

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4796949360/jobs/8533323696) [(plain)](https://github.com/rust-lang/rust/commit/35d13550d4def511d67af3ac6ba0cb69812a0f0d/checks/13004386052/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:8e5e7e5ab8b370d6c329ec480221332ada57f0ab)
Download action repository 'rust-lang/simpleinfra@master' (SHA:4d706a8d448f000965ed9e883febfb36661202fe)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: mingw-check-tidy
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180119 sha256=9fa76c45f3193f307e02ea67d1a48cfe7ef2b854a074b766938a88e046bc7887
  Stored in directory: /tmp/pip-ephem-wheel-cache-djp5cp8t/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
Successfully built 096c8bcd3848
Successfully tagged rust-ci:latest
Built container sha256:096c8bcd3848303878aec5f1a1113d5c9976260351fc7b181e6d62c1b55bd41b
Uploading finished image to https://ci-caches.rust-lang.org/docker/eee721fc1ad06e35509f22d2087fb4acdb0a4d378bec3a191179769e2503f17c7f8f4292d7013976df3a24502e822c590708d80a62f327ca4849564995e4deb7
upload failed: - to s3://rust-lang-ci-sccache2/docker/eee721fc1ad06e35509f22d2087fb4acdb0a4d378bec3a191179769e2503f17c7f8f4292d7013976df3a24502e822c590708d80a62f327ca4849564995e4deb7 Unable to locate credentials
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
fmt check
Diff in /checkout/compiler/rustc_middle/src/ty/subst.rs at line 67:
                 // Ensure we can use the tag bits.
 
                 let r: &WithCachedTypeInfo<ty::TyKind<'tcx>> = ty.0.pointer().0;
-                assert_eq!(mem::align_of_val::<WithCachedTypeInfo<ty::TyKind<'tcx>>>(r) & TAG_MASK, 0);
+                assert_eq!(
+                    mem::align_of_val::<WithCachedTypeInfo<ty::TyKind<'tcx>>>(r) & TAG_MASK,
+                );
+                );
                 (TYPE_TAG, r as *const WithCachedTypeInfo<ty::TyKind<'tcx>> as usize)
             }
             GenericArgKind::Const(ct) => {
Diff in /checkout/compiler/rustc_middle/src/ty/sty.rs at line 5:
 use crate::infer::canonical::Canonical;
 use crate::ty::subst::{GenericArg, InternalSubsts, SubstsRef};
 use crate::ty::visit::ValidateBoundVars;
-use crate::ty::{InferTy::*, HotTypeFlags};
 use crate::ty::{
     self, AdtDef, Discr, Term, Ty, TyCtxt, TypeFlags, TypeSuperVisitable, TypeVisitable,
     TypeVisitableExt, TypeVisitor,
Diff in /checkout/compiler/rustc_middle/src/ty/sty.rs at line 12:
 };
+use crate::ty::{HotTypeFlags, InferTy::*};
 use crate::ty::{List, ParamEnv};
 use hir::def::DefKind;
Diff in /checkout/compiler/rustc_middle/src/ty/sty.rs at line 1716:
     pub fn flags(self) -> TypeFlags {
         self.0.flags
     }
     }
-    
+
     #[inline(always)]
     pub fn hot_flags(self) -> HotTypeFlags {
         self.0.tag()
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_middle/src/ty/consts/int.rs" "/checkout/compiler/rustc_middle/src/ty/consts/kind.rs" "/checkout/compiler/rustc_middle/src/ty/query.rs" "/checkout/compiler/rustc_middle/src/ty/adjustment.rs" "/checkout/compiler/rustc_middle/src/ty/typeck_results.rs" "/checkout/compiler/rustc_middle/src/ty/subst.rs" "/checkout/compiler/rustc_middle/src/ty/list.rs" "/checkout/compiler/rustc_middle/src/ty/erase_regions.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

WaffleLapkin committed:
fmt

madsravn committed:
Updating Wake example to use new 'pin!' macro

WaffleLapkin commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit e1f392208507d7057a15a687323d99ec145c29f9 with merge a44290c85a010dc8d97a42dc2f35993b97f1805d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"e1f392208507d7057a15a687323d99ec145c29f9","merge_sha":"a44290c85a010dc8d97a42dc2f35993b97f1805d"} -->

bors committed:
Auto merge of #110325 - obeis:hir-analysis-migrate-diagnostics-4, r=davidtwco

Migrate `rustc_hir_analysis` to session diagnostic [Part 4]

Part 4: Finishing `check/mod.rs` file

r? `@compiler-errors`

onur-ozkan committed:
pass `unused_extern_crates` in `librustdoc::doctest::make_test`

Signed-off-by: ozkanonur <work@onurozkan.dev>

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4798201871/jobs/8538708926)
Build commit: a44290c85a010dc8d97a42dc2f35993b97f1805d (`a44290c85a010dc8d97a42dc2f35993b97f1805d`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4798201871/jobs/8538708926"},"merge_sha":"a44290c85a010dc8d97a42dc2f35993b97f1805d"} -->

rust-timer commented:
Queued a44290c85a010dc8d97a42dc2f35993b97f1805d with parent 999e6e5afb71d0fa6b5f67440278129aca12c67d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=999e6e5afb71d0fa6b5f67440278129aca12c67d&end=a44290c85a010dc8d97a42dc2f35993b97f1805d).
<!-- rust-timer: temporary -->

bors committed:
Auto merge of #110518 - loongarch-rs:update-linux-raw-sys, r=Mark-Simulacrum

Update linux-raw-sys to 0.3.4

To support LoongArch.

Changes:
```
Updating errno v0.2.8 -> v0.3.1
Updating is-terminal v0.4.4 -> v0.4.7
Updating linux-raw-sys v0.1.4 -> v0.3.4
Updating rustix v0.36.5 -> v0.37.7
Updating terminal_size v0.2.3 -> v0.2.6
  Adding windows-sys v0.48.0
```

The changes are generated by:

```bash
cargo update -p is-terminal -p terminal_size
```

cuishuang committed:
Remove repeated definite articles

Signed-off-by: cui fliter <imcusg@gmail.com>

rust-timer commented:
Finished benchmarking commit (a44290c85a010dc8d97a42dc2f35993b97f1805d): [comparison URL](https://perf.rust-lang.org/compare.html?start=999e6e5afb71d0fa6b5f67440278129aca12c67d&end=a44290c85a010dc8d97a42dc2f35993b97f1805d&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=999e6e5afb71d0fa6b5f67440278129aca12c67d&end=a44290c85a010dc8d97a42dc2f35993b97f1805d&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.1% | [0.3%, 3.2%] | 70    |
| Regressions ❌ <br /> (secondary)  | 1.7% | [0.1%, 4.9%] | 55    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.1% | [0.3%, 3.2%] | 70    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=999e6e5afb71d0fa6b5f67440278129aca12c67d&end=a44290c85a010dc8d97a42dc2f35993b97f1805d&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.2% | [-1.2%, -1.2%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.5% | [-2.5%, -2.5%] | 1     |
| All ❌✅ (primary)                 | -1.2% | [-1.2%, -1.2%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=999e6e5afb71d0fa6b5f67440278129aca12c67d&end=a44290c85a010dc8d97a42dc2f35993b97f1805d&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.8% | [0.7%, 3.7%] | 40    |
| Regressions ❌ <br /> (secondary)  | 3.1% | [1.7%, 4.9%] | 26    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.8% | [0.7%, 3.7%] | 40    |
</details>

WaffleLapkin commented:
`types_may_unify` just recursively calls `Ty::kind`, which suggests that `CopyTaggedPtr::{deref, pointer_raw}` may worsen codegen. They don't have `#[inline]`, so maybe that is the reason 🤔

cjgillot committed:
Move unstatisfaction check earlier.

WaffleLapkin committed:
Sprinkle some `#[inline]` in `rustc_data_structures::tagged_ptr`

WaffleLapkin committed:
Add `#[inline]` in `impl_tag`

fmease committed:
Add regression tests for const-generic IATs

clubby789 committed:
Migrate `rustc_passes` to translatable diagnostics

tamird committed:
simplify TrustedLen impls

Implement on FlattenCompat and delegate from Flatten and FlatMap.

matthiaskrgr committed:
Rollup merge of #110556 - kylematsuda:earlybinder-explicit-item-bounds, r=compiler-errors

Switch to `EarlyBinder` for `explicit_item_bounds`

Part of the work to finish https://github.com/rust-lang/rust/issues/105779.

This PR adds `EarlyBinder` to the return type of the `explicit_item_bounds` query and removes `bound_explicit_item_bounds`.

r? `@compiler-errors` (hope it's okay to request you, since you reviewed #110299 and #110498 :smiley:)

matthiaskrgr committed:
Rollup merge of #110615 - WaffleLapkin:impl_tag, r=cjgillot

Add `impl_tag!` macro to implement `Tag` for tagged pointer easily

r? `@Nilstrieb`

This should also lifts the need to think about safety from the callers (`impl_tag!` is robust (ish, see the macro issue)) and removes the possibility of making a "weird" `Tag` impl.

matthiaskrgr committed:
Rollup merge of #110649 - arlosi:fix_no_global_oom_handling, r=Mark-Simulacrum

Fix no_global_oom_handling build

`provide_sorted_batch` in core is incorrectly marked with `#[cfg(not(no_global_oom_handling))]` which prevents core from building with the cfg enabled.

Nothing in `core` allocates memory (including this function). The `cfg` gate is incorrect.

cc ``@dpaoliello``
r? ``@wesleywiser``

The cfg was added by #107191

matthiaskrgr committed:
Rollup merge of #110671 - compiler-errors:polarity, r=lcnr

Consider polarity in new solver

It's kinda ugly to have a polarity check in all of the builtin impls -- I guess I could consider the polarity at the top of assemble-builtin but that would require adding a polarity fn to `GoalKind`...

:shrug: putting this up just so i dont forget, since it's needed to bootstrap core during coherence (this alone does not allow core to bootstrap though, additional work is needed!)

r? ``@lcnr``

matthiaskrgr committed:
Rollup merge of #110783 - dtolnay:safeprint, r=petrochenkov

Fix ICE on --print=... i/o errors

`rustc --print=cfg > /dev/full`

**Before:** ICE (see below)

**After:** exits quietly with code 1.

```console
thread 'rustc' panicked at 'failed printing to stdout: No space left on device (os error 28)', library/std/src/io/stdio.rs:1019:9
stack backtrace:
   0:     0x7ffa84d68f33 - std::backtrace_rs::backtrace::libunwind::trace::h527d8d64d53ade2d
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7ffa84d68f33 - std::backtrace_rs::backtrace::trace_unsynchronized::hfb55b01517dd6379
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7ffa84d68f33 - std::sys_common::backtrace::_print_fmt::hd134e914eea0bd97
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7ffa84d68f33 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h1480db11ec399d77
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7ffa84dc9d4f - core::fmt::write::h67ec4c4171c92b26
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/core/src/fmt/mod.rs:1247:17
   5:     0x7ffa84d5bed1 - std::io::Write::write_fmt::h3b12aef0fff2463b
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/io/mod.rs:1712:15
   6:     0x7ffa84d68d45 - std::sys_common::backtrace::_print::h584400135abdbd51
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x7ffa84d68d45 - std::sys_common::backtrace::print::hce41d3c8bd91096b
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x7ffa84d6b84f - std::panicking::default_hook::{{closure}}::h2043b657a3791225
   9:     0x7ffa84d6b507 - std::panicking::default_hook::h99252b8d3dd5719c
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/panicking.rs:293:9
  10:     0x7ffa87fdf915 - <rustc_driver_impl[d30cd2737d9d343a]::DEFAULT_HOOK::{closure#0}::{closure#0} as core[cc19a662f3570270]::ops::function::FnOnce<(&core[cc19a662f3570270]::panic::panic_info::PanicInfo,)>>::call_once::{shim:vtable#0}
  11:     0x7ffa84d6c005 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h222a2b674b9f4762
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/alloc/src/boxed.rs:1976:9
  12:     0x7ffa84d6c005 - std::panicking::rust_panic_with_hook::h7f49b36bf7f8ff77
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/panicking.rs:704:13
  13:     0x7ffa84d6bd73 - std::panicking::begin_panic_handler::{{closure}}::haa23a7352589e31e
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/panicking.rs:595:13
  14:     0x7ffa84d69376 - std::sys_common::backtrace::__rust_end_short_backtrace::h3d0cf6e3c96e3fe9
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/sys_common/backtrace.rs:150:18
  15:     0x7ffa84d6ba72 - rust_begin_unwind
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/panicking.rs:584:5
  16:     0x7ffa84dc5fe3 - core::panicking::panic_fmt::hf4b4ea11e3fdb110
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/core/src/panicking.rs:67:14
  17:     0x7ffa84d5aa07 - std::io::stdio::print_to::h55760b9ede306280
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/io/stdio.rs:1019:9
  18:     0x7ffa84d5aa07 - std::io::stdio::_print::h93deb6099db33eab
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/io/stdio.rs:1095:5
  19:     0x7ffa871e163b - rustc_driver_impl[d30cd2737d9d343a]::print_crate_info
  20:     0x7ffa87fb8e01 - rustc_span[9551eaa044f53f4f]::set_source_map::<(), rustc_interface[c2b70c9b1dae0906]::interface::run_compiler<(), rustc_driver_impl[d30cd2737d9d343a]::run_compiler::{closure#0}>::{closure#0}::{closure#0}>
  21:     0x7ffa87fb6d59 - std[71a32ca0600a6a04]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[c2b70c9b1dae0906]::util::run_in_thread_pool_with_globals<rustc_interface[c2b70c9b1dae0906]::interface::run_compiler<(), rustc_driver_impl[d30cd2737d9d343a]::run_compiler::{closure#0}>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  22:     0x7ffa87fe25b4 - std[71a32ca0600a6a04]::panicking::try::<(), core[cc19a662f3570270]::panic::unwind_safe::AssertUnwindSafe<<std[71a32ca0600a6a04]::thread::Builder>::spawn_unchecked_<rustc_interface[c2b70c9b1dae0906]::util::run_in_thread_pool_with_globals<rustc_interface[c2b70c9b1dae0906]::interface::run_compiler<(), rustc_driver_impl[d30cd2737d9d343a]::run_compiler::{closure#0}>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1}::{closure#0}>>
  23:     0x7ffa87fb0cf1 - <<std[71a32ca0600a6a04]::thread::Builder>::spawn_unchecked_<rustc_interface[c2b70c9b1dae0906]::util::run_in_thread_pool_with_globals<rustc_interface[c2b70c9b1dae0906]::interface::run_compiler<(), rustc_driver_impl[d30cd2737d9d343a]::run_compiler::{closure#0}>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[cc19a662f3570270]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  24:     0x7ffa84d763e5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hf15d802f31f86225
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/alloc/src/boxed.rs:1962:9
  25:     0x7ffa84d763e5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hac564355b46c52d6
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/alloc/src/boxed.rs:1962:9
  26:     0x7ffa84d763e5 - std::sys::unix::thread::Thread::new::thread_start::h86fb3aedb7811f07
                               at /rustc/7f94b314cead7059a71a265a8b64905ef2511796/library/std/src/sys/unix/thread.rs:108:17
  27:     0x7ffa84a94b43 - start_thread
                               at ./nptl/pthread_create.c:442:8
  28:     0x7ffa84b26a00 - clone3
                               at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
  29:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.71.0-nightly (7f94b314c 2023-04-23) running on x86_64-unknown-linux-gnu

query stack during panic:
end of query stack
```

matthiaskrgr committed:
Rollup merge of #110796 - madsravn:wake-example, r=Mark-Simulacrum

Updating Wake example to use new 'pin!' macro

Closes: https://github.com/rust-lang/rust/issues/109965

I have already had this reviewed and approved here: https://github.com/rust-lang/rust/pull/110026 . But because I had some git issues and chose the "nuke it" option as my solution it didn't get merged. I nuked it too quickly. I am sorry for trouble of reviewing twice.

scottmcm committed:
Lower `intrinsics::offset` to `mir::BinOp::Offset`

They're semantically the same, so this means the backends don't need to handle the intrinsic and means fewer MIR basic blocks in pointer arithmetic code.

bors committed:
Auto merge of #110821 - matthiaskrgr:rollup-mzxrvw7, r=matthiaskrgr

Rollup of 6 pull requests

Successful merges:

 - #110556 (Switch to `EarlyBinder` for `explicit_item_bounds`)
 - #110615 (Add `impl_tag!` macro to implement `Tag` for tagged pointer easily)
 - #110649 (Fix no_global_oom_handling build)
 - #110671 (Consider polarity in new solver)
 - #110783 (Fix ICE on --print=... i/o errors)
 - #110796 (Updating Wake example to use new 'pin!' macro)

Failed merges:

r? `@ghost`
`@rustbot` modify labels: rollup

notriddle committed:
diagnostics: add test case for already-solved issue

Fixes #70082

bors committed:
Auto merge of #110811 - compiler-errors:vars-are-question-mark, r=WaffleLapkin

Use `?0` notation for ty/ct/int/float/region vars

Aligns the notation for infer vars that T-types and friends most often uses for inference variables with the notation in the compiler (which is kinda a sigil nightmare IMO: `_#`) by adopting `?0` style infer vars.

This mostly affects debug output since verbose infer vars shouldn't show up in user-facing places.

Does this need an MCP? It's debug output, so I'm thinking no, but happy to open one. :thinking:

r? types

weihanglo committed:
Update cargo

nnethercote committed:
Make some region folders a little stricter.

Because certain regions cannot occur in them.

bors committed:
Auto merge of #110834 - weihanglo:update-cargo, r=weihanglo

Update cargo

17 commits in de80432f04da61d98dcbbc1572598071718ccfd2..9e586fbd8b931494067144623b76c37d213b1ab6
2023-04-21 13:18:32 +0000 to 2023-04-25 22:09:11 +0000
- Update home dependency (rust-lang/cargo#12037)
- Warn instead of error in `cargo package` on empty `readme` or `license-file` in manifest (rust-lang/cargo#12036)
- Clarify documentation around test target setting. (rust-lang/cargo#12032)
- fix: apply `[env]` to target info discovery rustc (rust-lang/cargo#12029)
- CI: ensure intra links for all members are checked (rust-lang/cargo#12025)
- chore: make credential dependencies platform-specific (rust-lang/cargo#12027)
- CI: use `-p` to specify workspace members instead of `--manifest-path` (rust-lang/cargo#12024)
- ci: requires `test_gitoxide` and `lockfile` for both bors success and failure (rust-lang/cargo#12026)
- Update windows-sys (rust-lang/cargo#12021)
- Bump libc to 0.2.142 (rust-lang/cargo#12014)
- Update openssl-src to 111.25.3+1.1.1t (rust-lang/cargo#12005)
- Improve error message for empty dep (rust-lang/cargo#12001)
- Remove wrong url in benchsuite manifest. (rust-lang/cargo#12020)
- Bump versions of local crates (rust-lang/cargo#12019)
- Add the Win32_System_Console feature since it is used (rust-lang/cargo#12016)
- Update outdated crates.io URLs in publishing guide (rust-lang/cargo#12018)
- Allow named debuginfo options in Cargo.toml (rust-lang/cargo#11958)

r? `@ghost`

bors committed:
Auto merge of #110634 - saethlin:pointy-decoder, r=cjgillot

Rewrite MemDecoder around pointers not a slice

This is basically https://github.com/rust-lang/rust/pull/109910 but I'm being a lot more aggressive. The pointer-based structure means that it makes a lot more sense to absorb more complexity into `MemDecoder`, most of the diff is just complexity moving from one place to another.

The primary argument for this structure is that we only incur a single bounds check when doing multi-byte reads from a `MemDecoder`. With the slice-based implementation we need to do those with `data[position..position + len]` , which needs to account for `position + len` wrapping. It would be possible to dodge the first bounds check if we stored a slice that starts at `position`, but that would require updating the pointer and length on every read.

This PR also embeds the failure path in a separate function, which means that this PR should subsume all the perf wins observed in https://github.com/rust-lang/rust/pull/109867.

bors committed:
Auto merge of #110784 - cuviper:build-gcc-https, r=jyn514

Revert "Download the GCC sources insecurely"

This reverts commit 3da037f82988b8b3aca2ce13c5c81ba975923cab.

This workaround was added after TLS problems with Debian 6 were noted in <https://github.com/rust-lang/rust/pull/86586#issuecomment-868355356>, but we should be well past that since #95026, where our oldest images are now based on CentOS 7.

jyn514 committed:
Rollup merge of #108416 - pat-nel87:Issue-107957-black_box_docs, r=jyn514

black_box doc corrections for clarification - Issue #107957

Made a complete pass through the docs to help resolve https://github.com/rust-lang/rust/issues/107957

No code changes, just documentation

`@rustbot` label +T-libs-api -T-libs

jyn514 committed:
Rollup merge of #110266 - tgross35:try-exists-wording, r=jyn514

Update documentation wording on path 'try_exists' functions

Just eliminate the quadruple negation in `doesn't silently ignore errors unrelated to ... not existing.`

jyn514 committed:
Rollup merge of #109379 - flba-eb:108596_fixtest_sigpipe, r=jyn514

Replace `yes` command by `while-echo` in test `tests/ui/process/process-sigpipe.rs`

The `yes` command is not available on all platforms.

Fixes #108596.

Inviting `@mvf` as he contributed to this patch. Thanks! This issue has been discussed in https://github.com/rust-lang/rust/pull/106673 but was moved to #108596 to get going.

CC `@gh-tr`

r? `@workingjubilee`
`@rustbot` label +O-neutrino

Notes about the comments https://github.com/rust-lang/rust/pull/106673#discussion_r1117324265:
- The `echo` command is `/proc/boot/echo` (not built-in)
- `/bin/sh` is a symlink to `/proc/boot/ksh`
```sh
# ls -l /bin/sh /proc/boot/ksh /proc/boot/echo
lrwxrwxrwx   1 root      root             14 Mar 20 07:52 /bin/sh -> /proc/boot/ksh
-r-xr-xr-x   1 root      root           9390 Sep 12  2022 /proc/boot/echo
-r-xr-xr-x   1 root      root         308114 Sep 12  2022 /proc/boot/ksh
```

jyn514 committed:
Rollup merge of #110418 - jsoref:spelling-rustdoc, r=jyn514

Spelling rustdoc

Split per https://github.com/rust-lang/rust/pull/110392#issuecomment-1510148682

jyn514 committed:
Rollup merge of #110329 - aDotInTheVoid:json-inline-again, r=jyn514

Improve tests for #110138

These should live in rustdoc-json, not rustdoc-ui, so we can run assertions, and not just check there's no ICE

CC #100515, as we never document this suite

r? rustdoc

jyn514 committed:
Rollup merge of #110587 - tomaka:fix-109727, r=jyn514

Fix `std` compilation error for wasi+atomics

Fix https://github.com/rust-lang/rust/issues/109727

It seems that the `unsupported/once.rs` module isn't meant to exist at the same time as the `futex` module, as they have conflicting definitions.

I've solved this by defining the `once` module only if `not(target_feature = "atomics")`.
The `wasm32-unknown-unknown` target [similarly only defines the `once` module if `not(target_feature = "atomics")`](https://github.com/tomaka/rust/blob/01c4f319276da912dd2be768ae0ce9857ad6bb63/library/std/src/sys/wasm/mod.rs#L69-L70).

As show in [this block of code](https://github.com/tomaka/rust/blob/01c4f319276da912dd2be768ae0ce9857ad6bb63/library/std/src/sys_common/once/mod.rs#L10-L34), the `sys::once` module doesn't need to exist if `all(target_arch = "wasm32", target_feature = "atomics")`.

jyn514 committed:
Rollup merge of #110594 - infdahai:cfg_chore, r=jyn514

`rustc --help` add `--cfg` SPEC declaration.

1. fixes #110462
2. add spec arguments based on https://doc.rust-lang.org/reference/conditional-compilation.html

jyn514 committed:
Rollup merge of #110817 - fmease:fix-109759, r=compiler-errors

Add regression tests for const-generic inherent associated types

Fixes #109759.
The tests are no longer failing since #96840 which was merged recently (#109410 is no longer necessary for them).

`@rustbot` label F-inherent_associated_types

jyn514 committed:
Rollup merge of #110792 - rylev:standard-macos-ci, r=pietroalbini

Use the standard macOS CI runner

This moves back to the standard macOS CI runner instead of the large runners. The infra team is discussing whether the large runners are worth the cost at our heavy usage and whether those credits are better spent elsewhere. This will slow down CI build times unfortunately, but while we figure out our CI spending budget, that's a price we'll unfortunately have to pay.

r? `@pietroalbini`

bors committed:
Auto merge of #110839 - jyn514:rollup-uikilwm, r=jyn514

Rollup of 9 pull requests

Successful merges:

 - #108416 (black_box doc corrections for clarification - Issue #107957)
 - #109379 (Replace `yes` command by `while-echo` in test `tests/ui/process/process-sigpipe.rs`)
 - #110266 (Update documentation wording on path 'try_exists' functions)
 - #110329 (Improve tests for #110138)
 - #110418 (Spelling rustdoc)
 - #110587 (Fix `std` compilation error for wasi+atomics)
 - #110594 (`rustc --help` add `--cfg` SPEC declaration.)
 - #110792 (Use the standard macOS CI runner)
 - #110817 (Add regression tests for const-generic inherent associated types)

Failed merges:

r? `@ghost`
`@rustbot` modify labels: rollup

bors committed:
Auto merge of #97368 - tmandry:coverage-underflow, r=jyn514

coverage: Don't underflow column number

I noticed this when running coverage on a debug build of rustc. There
may be other places that do this but I'm just fixing the one I hit.

r? `@wesleywiser` `@richkadel`

aDotInTheVoid committed:
rustdoc-json: Time serialization.

bors committed:
Auto merge of #110822 - scottmcm:lower-offset-to-mir, r=compiler-errors

Lower `intrinsics::offset` to `mir::BinOp::Offset`

They're [semantically the same](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.Rvalue.html#variant.BinaryOp), so this means the backends don't need to handle the intrinsic and means fewer MIR basic blocks in pointer arithmetic code.

matthiaskrgr committed:
Rollup merge of #108760 - clubby789:autolintstuff, r=wesleywiser

Add lint to deny diagnostics composed of static strings

r? ghost

I'm hoping to have a lint that semi-automatically converts simple diagnostics such as `struct_span_err(span, "msg").help("msg").span_note(span2, "msg").emit()` to typed session diagnostics. It's quite hacky and not entirely working because of problems with `x fix` but should hopefully help reduce some of the work.
I'm going to start trying to apply what I can from this, but opening this as a draft in case anyone wants to develop on it.

cc #100717

matthiaskrgr committed:
Rollup merge of #109444 - WaffleLapkin:undeprecate_todos, r=jyn514

Change tidy error message for TODOs

Blocked on #109440 (first few commits are from where)

IMO "deprecated" doesn't really explain anything, I've tried to highlight the actual reason we error on TODOs. The message is not at all perfect, maybe someone has ideas how to phrase it better?

matthiaskrgr committed:
Rollup merge of #110550 - compiler-errors:deref-on-binop-rhs, r=wesleywiser

Suggest deref on comparison binop RHS even if type is not Copy

Fixes #110500

matthiaskrgr committed:
Rollup merge of #110419 - jsoref:spelling-library, r=jyn514

Spelling library

Split per https://github.com/rust-lang/rust/pull/110392

I can squash once people are happy w/ the changes. It's really uncommon for large sets of changes to be perfectly acceptable w/o at least some changes.

I probably won't have time to respond until tomorrow or the next day

matthiaskrgr committed:
Rollup merge of #110798 - ozkanonur:rustdoc-unused-extern-crates, r=jyn514

pass `unused_extern_crates` in `librustdoc::doctest::make_test`

blocker for https://github.com/rust-lang/rust/pull/106621

matthiaskrgr committed:
Rollup merge of #110641 - GuillaumeGomez:rustdoc-in-doc-settings, r=notriddle

Add new rustdoc book chapter to describe in-doc settings

Fixes #55165.

I continue going through old rustdoc issues. This one made a lot of sense so decided to add the missing chapter.

r? ``@notriddle``

matthiaskrgr committed:
Rollup merge of #110819 - tamird:flattencompat-trustedlen, r=the8472

simplify TrustedLen impls

Implement on FlattenCompat and delegate from Flatten and FlatMap.

/cc ``@the8472``

matthiaskrgr committed:
Rollup merge of #110825 - notriddle:notriddle/issue-70082, r=compiler-errors

diagnostics: add test case for already-solved issue

Fixes #70082

matthiaskrgr committed:
Rollup merge of #110835 - nnethercote:strict-region-folders-2, r=compiler-errors

Make some region folders a little stricter.

Because certain regions cannot occur in them.

r? ``@compiler-errors``

matthiaskrgr committed:
Rollup merge of #110847 - aDotInTheVoid:rdj-time-serialization, r=GuillaumeGomez

rustdoc-json: Time serialization.

This lets us know how much time is spent in json specific code, and how much performance we could gain by using a different serialization format.

For aws-sdk-ec2, it takes 0.7s, out of a 43s build (of which 6s is spent in json specific code), and makes a 173M json file

```
$ cargo +rust_stage2 rustdoc -p aws-sdk-ec2 -- -Zunstable-options -w json -Ztime-passes
 Documenting aws-sdk-ec2 v0.26.0 (/home/gh-aDotInTheVoid/aws-sdk-rust/sdk/ec2)
time:   5.229; rss:   64MB ->  891MB ( +827MB)  expand_crate
time:   5.230; rss:   61MB ->  891MB ( +830MB)  macro_expand_crate
time:   0.256; rss:  891MB ->  891MB (   +0MB)  AST_validation
time:   0.025; rss:  891MB ->  891MB (   +1MB)  finalize_imports
time:   0.093; rss:  891MB ->  891MB (   +0MB)  compute_effective_visibilities
time:   0.122; rss:  891MB ->  891MB (   +0MB)  finalize_macro_resolutions
time:   2.442; rss:  891MB -> 1188MB ( +297MB)  late_resolve_crate
time:   0.120; rss: 1188MB -> 1190MB (   +2MB)  resolve_check_unused
time:   0.211; rss: 1190MB -> 1190MB (   +0MB)  resolve_postprocess
time:   3.017; rss:  891MB -> 1190MB ( +299MB)  resolve_crate
time:   8.520; rss:   61MB -> 1181MB (+1120MB)  prepare_outputs
time:   0.152; rss: 1181MB -> 1181MB (   +0MB)  complete_gated_feature_checking
time:   0.539; rss: 1633MB -> 1632MB (   -1MB)  drop_ast
time:  15.492; rss:   58MB -> 1621MB (+1563MB)  type_collecting
time:   1.503; rss: 1621MB -> 1705MB (  +84MB)  item_types_checking
time:   1.274; rss: 1705MB -> 1726MB (  +21MB)  crate_lints
time:   1.275; rss: 1705MB -> 1726MB (  +21MB)  missing_docs
time:   0.281; rss: 1726MB -> 1726MB (   +0MB)  check_mod_attrs
time:   0.433; rss: 1744MB -> 1750MB (   +6MB)  clean_crate
time:  11.581; rss: 1750MB -> 2107MB ( +357MB)  collect_synthetic_impls
time:   0.019; rss: 2107MB -> 2107MB (   +0MB)  collect_items_for_trait_impls
time:  12.588; rss: 1750MB -> 2139MB ( +389MB)  collect-trait-impls
time:   0.197; rss: 2139MB -> 2139MB (   +0MB)  check_doc_test_visibility
time:   0.281; rss: 2139MB -> 2150MB (  +11MB)  strip-hidden
time:   0.260; rss: 2150MB -> 2150MB (   +0MB)  strip-private
warning: unresolved link to `date`
 --> sdk/ec2/src/client/describe_instances.rs:7:11184
  |
7 | ...te of the instance (for example, shows "User Initiated [date]" when you stop or terminate the instance). Similar to the state-reason-code filter.<...
  |                                                            ^^^^ no item named `date` in scope
  |
  = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
  = note: `#[warn(rustdoc::broken_intra_doc_links)]` on by default

time:   0.363; rss: 2150MB -> 2150MB (   +0MB)  collect-intra-doc-links
time:   0.946; rss: 2150MB -> 2150MB (   +0MB)  propagate-doc-cfg
time:   0.494; rss: 2150MB -> 2152MB (   +2MB)  run-lints
time:   0.658; rss: 2152MB -> 2163MB (  +11MB)  create_format_cache
time:  34.818; rss:   58MB -> 2163MB (+2105MB)  run_global_ctxt
time:   0.016; rss: 2163MB -> 2163MB (   +0MB)  create_renderer(json)
time:   0.723; rss: 2780MB -> 2780MB (   +0MB)  rustdoc_json_serialization
time:   2.216; rss: 2399MB -> 2599MB ( +199MB)  renderer_after_krate(json)
time:   6.639; rss: 2163MB -> 2417MB ( +254MB)  render_json
time:   0.312; rss: 2417MB -> 2071MB ( -346MB)  free_global_ctxt
warning: `aws-sdk-ec2` (lib doc) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 43.06s
```

[(Zulip Discussion)](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/JSON.20metadata.20speed/near/352783145)

r? `@GuillaumeGomez`

bors committed:
Auto merge of #110852 - matthiaskrgr:rollup-jz3eosr, r=matthiaskrgr

Rollup of 10 pull requests

Successful merges:

 - #108760 (Add lint to deny diagnostics composed of static strings)
 - #109444 (Change tidy error message for TODOs)
 - #110419 (Spelling library)
 - #110550 (Suggest deref on comparison binop RHS even if type is not Copy)
 - #110641 (Add new rustdoc book chapter to describe in-doc settings)
 - #110798 (pass `unused_extern_crates` in `librustdoc::doctest::make_test`)
 - #110819 (simplify TrustedLen impls)
 - #110825 (diagnostics: add test case for already-solved issue)
 - #110835 (Make some region folders a little stricter.)
 - #110847 (rustdoc-json: Time serialization.)

Failed merges:

r? `@ghost`
`@rustbot` modify labels: rollup

m-ou-se committed:
Restructure and rename thread local things in std.

m-ou-se committed:
Remove libstd_thread_internals from unstable book.

m-ou-se committed:
Update tests.

compiler-errors committed:
IntoFuture::into_future is no longer unstable

compiler-errors committed:
Make method-not-found-generic-arg-elision.rs error message not path dependent

bors committed:
Auto merge of #110861 - m-ou-se:thread-local-restructure, r=workingjubilee

Restructure and rename std thread_local internals to make it less of a maze

Every time I try to work on std's thread local internals, it feels like I'm trying to navigate a confusing maze made of macros, deeply nested modules, and types with multiple names/aliases. Time to clean it up a bit.

This PR:

- Exports `Key` with its own name (`Key`), instead of `__LocalKeyInner`
- Uses `pub macro` to put `__thread_local_inner` into a (unstable, hidden) module, removing `#[macro_export]`, removing it from the crate root.
- Removes the `__` from `__thread_local_inner`.
- Removes a few unnecessary `allow_internal_unstable` features from the macros
- Removes the `libstd_thread_internals` feature. (Merged with `thread_local_internals`.)
    - And removes it from the unstable book
- Gets rid of the deeply nested modules for the `Key` definitions (`mod fast` / `mod os` / `mod statik`).
- Turns a `#[cfg]` mess into a single `cfg_if`, now that there's no `#[macro_export]` anymore that breaks with `cfg_if`.
- Simplifies the `cfg_if` conditions to not repeat the conditions.
- Removes useless `normalize-stderr-test`, which were left over from when the `Key` types had different names on different platforms.
- Removes a seemingly unnecessary `realstd` re-export on `cfg(test)`.

This PR changes nothing about the thread local implementation. That's for a later PR. (Which should hopefully be easier once all this stuff is a bit cleaned up.)

Jules-Bertholet committed:
Nicer ICE for #67981

bors committed:
Auto merge of #110562 - ComputerDruid:riscv, r=tmandry

Add definitions for riscv64gc-unknown-fuchsia

To compile, also requires a libc update with https://github.com/rust-lang/libc/pull/3204

bors committed:
Auto merge of #110728 - cjgillot:no-false-optes, r=oli-obk

Do not bother optimizing impossible functions.

This is currently checked by `ConstProp`, but I see no reason to restrict it to ConstProp only.

BoxyUwU committed:
rename `needs_subst` to `has_param`

BoxyUwU committed:
rename `needs_infer` to `has_infer`

bors committed:
Auto merge of #110882 - BoxyUwU:rename-some-ty-flags, r=compiler-errors

rename `NEEDS_SUBST` and `NEEDS_INFER`

implements rust-lang/compiler-team#617

matthiaskrgr committed:
Rollup merge of #110426 - calebcartwright:style-let-else-examples, r=compiler-errors

docs(style): add more let-else examples

Adding a few more examples for increased clarity based on subsequent discussion that arose after implementation work began.

Will need approval from t-style, but also cc ```@ytmimi``` to make sure these examples don't contradict any examples we've been incorporating as part of the rustfmt implementation

matthiaskrgr committed:
Rollup merge of #110804 - cuishuang:master, r=jyn514

Remove repeated definite articles

matthiaskrgr committed:
Rollup merge of #110814 - WaffleLapkin:sprinkle_#inline, r=Nilstrieb

Sprinkle some `#[inline]` in `rustc_data_structures::tagged_ptr`

This is based on `nm --demangle (rustc +a --print sysroot)/lib/librustc_driver-*.so | rg CopyTaggedPtr` which shows many methods that should probably be inlined. May fix the regression in https://github.com/rust-lang/rust/pull/110795.

r? ```@Nilstrieb```

matthiaskrgr committed:
Rollup merge of #110816 - clubby789:rustc-passes-diagnostics, r=compiler-errors

Migrate `rustc_passes` to translatable diagnostics

cc #100717

matthiaskrgr committed:
Rollup merge of #110866 - compiler-errors:test, r=jyn514

Make `method-not-found-generic-arg-elision.rs` error message not path dependent

Every time I bless `tests/ui/methods/method-not-found-generic-arg-elision.rs`, I get some nonsense "type is too long" + "written to disk" that shows up and have to manually revert because the combination of my rustc repo path + the UI test directory hits the length limit for printing types spilling to disk (since this happens before UI test path sanitization).

The fact that we use a closure in this test doesn't have to do with the UI test, so just box the closure to make the type name smaller and not path dependent.

matthiaskrgr committed:
Rollup merge of #110864 - compiler-errors:into-future-stable, r=jackh726

`IntoFuture::into_future` is no longer unstable

We don't need to gate the `IntoFuture::into_future` call in `.await` lowering anymore.

``@bors`` rollup

matthiaskrgr committed:
Rollup merge of #110872 - Jules-Bertholet:err-67981, r=wesleywiser

Nicer ICE for #67981

Provides a slightly nicer ICE for #67981, documenting the problem. A proper fix will be necessary before `#![feature(unsized_fn_params)]` can be stabilized.

The problem is that the design of the `"rust-call"` ABI is fundamentally not compatible with `unsized_fn_params`. `"rust-call"` functions need to collect their arguments into a tuple, but if the arguments are not `Sized`, said tuple is potentially not even a valid type—and if it is, it requires `alloca` to create.

``@rustbot`` label +A-abi +A-codegen +F-unboxed_closures +F-unsized_fn_params

bors committed:
Auto merge of #110896 - matthiaskrgr:rollup-h8fetzd, r=matthiaskrgr

Rollup of 7 pull requests

Successful merges:

 - #110426 (docs(style): add more let-else examples)
 - #110804 (Remove repeated definite articles)
 - #110814 (Sprinkle some `#[inline]` in `rustc_data_structures::tagged_ptr`)
 - #110816 (Migrate `rustc_passes` to translatable diagnostics)
 - #110864 (`IntoFuture::into_future` is no longer unstable)
 - #110866 (Make `method-not-found-generic-arg-elision.rs` error message not path dependent)
 - #110872 (Nicer ICE for #67981)

Failed merges:

r? `@ghost`
`@rustbot` modify labels: rollup

WaffleLapkin committed:
akncwe;ue3v;yer;yr:b

WaffleLapkin committed:
?

WaffleLapkin committed:
hello??

jyn514 commented:
![Screenshot 2023-04-27 at 1 59 40 PM](https://user-images.githubusercontent.com/23638587/234965018-6b7bcd93-3d3e-475d-881e-3af04c8de707.png)
uhhhhh
